### ISSUE TITLE
Is it safe to have Wqkv be float16?

### ISSUE THREAD

**Issue by vedantroy** [HazyResearch/flash-attention#49]
I notice that the QKV matrix is in fp16. Normally, I just use `torch.cuda.amp.autocast` to handle converting fp32 operations into fp16, so I'm not actually sure if:
1. autocast converts the QKV weight matrix into fp16
2. If the above is false, will the results be very different for fp16 attention vs fp32 attention / is something I need to take into consideration when training.

I'm asking, because I implemented fp16 attention to make a comparison against flash attention (to determine if the results are similar), and I got NaNs, which makes me wonder if fp16 attention is safe:

```
class Attention(nn.Module):
    def __init__(self, dim: int, heads: int):
        super().__init__()
        self.heads = heads
        head_dim = dim // heads
        self.scale = head_dim**-0.5

        self.softmax= nn.Softmax(dim=-1)
        self.to_qkv = nn.Linear(dim, dim * 3, bias=False, dtype=torch.float16)

    def forward(self, x):
        b, n_patches, dim = x.shape
        qkv = self.to_qkv(x).chunk(3, dim=-1)

        assert qkv[0].shape == x.shape

        q, k, v = map(lambda t: rearrange(t, "b n (h d) -> b h n d", h=self.heads), qkv)

        assert q.shape == (b, self.heads, n_patches, dim // self.heads)

        attn_matrix = (q @ k.transpose(-1, -2)) * self.scale

        assert attn_matrix.shape[-1] == attn_matrix.shape[-2]
        assert attn_matrix.shape[-1] == n_patches

        normalized_scores = self.softmax(attn_matrix)

        reaveraged_values = normalized_scores @ v
        reaveraged_values = rearrange(reaveraged_values, "b h n d -> b n (h d)")

        return reaveraged_values
```

**Edit:** Up-casting before applying attention does help, but I still have:
- max diff = 0.03125
- mean diff = 0.0001729

**Comment by tridao** [comment]
With FlashAttention we've trained BERT & GPT2 in fp16/bf16 just fine.
We have [tests](https://github.com/HazyResearch/flash-attention/blob/2ed471ecc450801cbe30be2169936a7fe69058c4/tests/test_flash_attn.py#L520) the check the numerical error of FlashAttention and a reference implementation, and the numerical error is similar to the Pytorch implementation.

Standard attention fp16 may or may not be safe, depends on how large the inputs (q, k, v) are and how you're implementing it. One place where it might NaN out is (Q @ K.T), which might produce entries larger than 64k (limit of fp16), before multiplying by self.scale. You can multiply Q by self.scale first before multiply by K.T.
There are [several](https://crfm.stanford.edu/2021/08/26/mistral.html) [places](https://github.com/bigscience-workshop/bigscience/blob/master/train/tr8-104B-wide/chronicles.md) you can read about numerical stability of attention.

**Comment by vedantroy** [comment]
> With FlashAttention we've trained BERT & GPT2 in fp16/bf16 just fine. We have [tests](https://github.com/HazyResearch/flash-attention/blob/2ed471ecc450801cbe30be2169936a7fe69058c4/tests/test_flash_attn.py#L520) the check the numerical error of FlashAttention and a reference implementation, and the numerical error is similar to the Pytorch implementation.
> 
> Standard attention fp16 may or may not be safe, depends on how large the inputs (q, k, v) are and how you're implementing it. One place where it might NaN out is (Q @ K.T), which might produce entries larger than 64k (limit of fp16), before multiplying by self.scale. You can multiply Q by self.scale first before multiply by K.T. There are [several](https://crfm.stanford.edu/2021/08/26/mistral.html) [places](https://github.com/bigscience-workshop/bigscience/blob/master/train/tr8-104B-wide/chronicles.md) you can read about numerical stability of attention.

Thanks for the reference. Do you have any tips on how to integrate this with Pytorch's amp? It seems like flash attention wants the QKV matrix to be in fp16. However, when training with Pytorch this causes the exception: "Attempting to unscale FP16 gradients.", which makes sense because I guess Pytorch expects the model's weights to be in FP32 (or something like that?)

**Comment by tridao** [comment]
You don't have to do any manual casting if you're running with Pytorch AMP.
The matrix W_qkv is in fp32 but AMP will cast the output of the matmul (qkv) to fp16/bf16, which then get passed to FlashAttention.

**Comment by vedantroy** [comment]
Amazing! got it working and it managed to 10x my batch size, which is super useful since I'm doing contrastive loss. Thanks!
